File: scripts/flame_graph/flamegraph.js
/*
 * -------------------------------------------------------------------------
 * This file is part of the MindStudio project.
 * Copyright (c) 2026 Huawei Technologies Co.,Ltd.
 *
 * MindStudio is licensed under Mulan PSL v2.
 * You can use this software according to the terms and conditions of the Mulan PSL v2.
 * You may obtain a copy of Mulan PSL v2 at:
 *
 *          http://license.coscl.org.cn/MulanPSL2
 *
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
 * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
 * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
 * See the Mulan PSL v2 for more details.
 * -------------------------------------------------------------------------
 *
 * PyTorch Flame Graph — Interactive Canvas Renderer
 *
 * Reads flame graph data from the inline <script id="flamegraph-data"> element
 * in the host HTML and renders an interactive, zoomable flame graph on <canvas>.
 */
(() => {
  // Show a friendly message instead of a blank page when initialization fails.
  const showFatalError = (message) => {
    document.body.innerHTML =
      `<div style="padding:24px;font-family:monospace;color:#d4d4d4;background:#1e1e1e;">`
      + `Failed to load flame graph: ${message}</div>`;
  };

  const dataEl = document.getElementById('flamegraph-data');
  if (!dataEl) {
    showFatalError('flame graph data element not found.');
    throw new Error('flamegraph-data element missing');
  }

  let DATA;
  try {
    DATA = JSON.parse(dataEl.textContent);
  } catch (err) {
    showFatalError('flame graph data is not valid JSON.');
    throw err;
  }

  if (!DATA || !DATA.flamegraph) {
    showFatalError('flame graph data is empty or malformed.');
    throw new Error('DATA.flamegraph missing');
  }

  const CATEGORY_COLORS = {
    root: '#5d4037',
    python: '#ef6c00',
    python_framework: '#f9a825',
    cann: '#c62828',
    unknown: '#8d6e63',
  };
  const ROW_HEIGHT = 22;
  const FONT_SIZE = 12;
  const MIN_LABEL_WIDTH = 60;
  const ANCESTOR_OPACITY = 0.25;
  const MAX_API_STACK_DEPTH = 1000;

  const canvas = document.getElementById('canvas');
  const ctx = canvas.getContext('2d');
  const container = document.getElementById('container');
  const tooltip = document.getElementById('tooltip');
  const searchInput = document.getElementById('searchInput');
  const searchCount = document.getElementById('searchCount');
  const resetBtn = document.getElementById('resetBtn');
  const resetSearchBtn = document.getElementById('resetSearchBtn');
  const threadDropdown = document.getElementById('threadDropdown');
  const threadDropdownToggle = document.getElementById('threadDropdownToggle');
  const threadDropdownMenu = document.getElementById('threadDropdownMenu');

  const metadata = DATA.metadata || {};
  const allRoot = DATA.flamegraph;
  let activeRoot = allRoot;
  let totalDuration = activeRoot.value || metadata.total_duration_us || 0;
  let searchText = '';
  let focusPath = [activeRoot];
  let layoutCache = null;
  let hoveredNode = null;
  let renderScheduled = false;
  let maxDepthWarningShown = false;

  const showMaxDepthWarning = () => {
    if (maxDepthWarningShown) return;
    maxDepthWarningShown = true;
    console.warn(`The flame graph exceeds the max API call stack depth (${MAX_API_STACK_DEPTH}); deeper frames will not be rendered.`);
  };

  const focused = () => focusPath[focusPath.length - 1];

  const scheduleRender = () => {
    if (renderScheduled) return;
    renderScheduled = true;
    requestAnimationFrame(() => {
      renderScheduled = false;
      render();
    });
  };

  const addThreadDropdownItem = (value, label) => {
    const item = document.createElement('div');
    item.className = 'dropdown-item';
    item.dataset.value = value;
    item.textContent = label;
    item.addEventListener('click', () => {
      setActiveThread(value);
      threadDropdown.classList.remove('open');
    });
    threadDropdownMenu.appendChild(item);
  };

  const populateThreadDropdown = () => {
    threadDropdownMenu.innerHTML = '';
    addThreadDropdownItem('__all__', 'All');
    if (!allRoot.children) return;

    const threads = [];
    for (const child of allRoot.children) {
      if (!child.name.startsWith('thread_')) continue;
      const match = child.name.match(/^thread_(\d+)/);
      const tid = match ? parseInt(match[1], 10) : 0;
      threads.push({ name: child.name, tid });
    }

    threads.sort((a, b) => a.tid - b.tid);
    for (const thread of threads) {
      addThreadDropdownItem(thread.name, thread.name);
    }
  };

  const updateThreadDropdownActive = (value) => {
    const items = threadDropdownMenu.querySelectorAll('.dropdown-item');
    for (const item of items) {
      item.classList.toggle('active', item.dataset.value === value);
    }
  };

  const setActiveThread = (threadName) => {
    activeRoot = allRoot;
    if (threadName !== '__all__' && allRoot.children) {
      const thread = allRoot.children.find((child) => child.name === threadName);
      if (thread) activeRoot = thread;
    }

    totalDuration = activeRoot.value || metadata.total_duration_us || 0;
    threadDropdownToggle.textContent = threadName === '__all__' ? 'All' : threadName;
    updateThreadDropdownActive(threadName);
    focusPath = [activeRoot];
    hoveredNode = null;
    tooltip.style.display = 'none';
    layoutCache = null;
    resize(true);
  };

  // Store parent links so zoom paths can be rebuilt without tree-wide searches.
  const initParentLinks = (root) => {
    root._parent = null;

    const stack = [root];
    while (stack.length > 0) {
      const node = stack.pop();
      if (!node.children) continue;

      for (const child of node.children) {
        child._parent = node;
        stack.push(child);
      }
    }
  };

  const isAncestor = (node) => focusPath.slice(0, -1).includes(node);

  // Recreate the backing canvas at device-pixel resolution after viewport changes.
  const resize = (alignBottom) => {
    let width = container.clientWidth;
    if (container.scrollHeight > container.clientHeight) {
      width -= container.offsetWidth - container.clientWidth;
    }

    const depth = getMaxDepth(activeRoot);
    const height = Math.max(depth * ROW_HEIGHT + 20, container.clientHeight, 300);
    const dpr = window.devicePixelRatio || 1;

    canvas.width = width * dpr;
    canvas.height = height * dpr;
    canvas.style.width = `${width}px`;
    canvas.style.height = `${height}px`;
    ctx.setTransform(dpr, 0, 0, dpr, 0, 0);
    layoutCache = null;
    render();

    if (alignBottom) {
      container.scrollTop = Math.max(0, container.scrollHeight - container.clientHeight);
    }
  };

  // Use an explicit stack to avoid recursion limits on very deep flame graphs.
  const getMaxDepth = (root) => {
    let max = 1;
    const stack = [{ node: root, depth: 1 }];

    while (stack.length > 0) {
      const { node, depth } = stack.pop();
      if (depth > max) max = depth;
      if (!node.children) continue;

      for (const child of node.children) {
        stack.push({ node: child, depth: depth + 1 });
      }
    }

    return max;
  };

  // Ancestors in the focus path keep full width; descendants are proportional to duration.
  const layoutTree = (node, x, width, depth) => {
    node._x = x;
    node._w = width;
    node._d = depth;

    if (depth >= MAX_API_STACK_DEPTH) {
      if (node.children && node.children.length > 0) showMaxDepthWarning();
      return;
    }

    for (let i = 0; i < focusPath.length - 1; i++) {
      if (focusPath[i] === node) {
        layoutTree(focusPath[i + 1], x, width, depth + 1);
        return;
      }
    }

    if (!node.children) return;

    let childX = x;
    const totalChildValue = node.children.reduce((sum, child) => sum + child.value, 0);
    const layoutValue = Math.max(node.value, totalChildValue);

    for (const child of node.children) {
      const childWidth = layoutValue > 0 ? (child.value / layoutValue) * width : 0;
      layoutTree(child, childX, childWidth, depth + 1);
      childX += childWidth;
    }
  };

  const getLayout = () => {
    if (layoutCache) return;
    layoutTree(activeRoot, 0, 1.0, 0);
    layoutCache = true;
  };

  // Add a stable name-based variation so adjacent frames in the same category are distinguishable.
  const getColor = (node) => {
    const base = CATEGORY_COLORS[node.category] || CATEGORY_COLORS.unknown;
    let hash = 0;
    for (const char of node.name) {
      hash = ((hash << 5) - hash + char.charCodeAt(0)) | 0;
    }

    const variation = (Math.abs(hash) % 30) - 15;
    const clamp = (value) => Math.max(0, Math.min(255, value));
    const r = clamp(parseInt(base.slice(1, 3), 16) + variation);
    const g = clamp(parseInt(base.slice(3, 5), 16) + variation);
    const b = clamp(parseInt(base.slice(5, 7), 16) + variation);
    return `rgb(${r},${g},${b})`;
  };

  const matchesSearch = (node) => {
    if (!searchText) return true;
    return node.name.toLowerCase().includes(searchText.toLowerCase());
  };

  const countMatches = (node) => {
    if (!searchText) return { total: 0, matched: 0 };

    let total = 0;
    let matched = 0;
    const walk = (current) => {
      total++;
      if (matchesSearch(current)) matched++;
      if (current.children) {
        for (const child of current.children) walk(child);
      }
    };

    walk(node);
    return { total, matched };
  };

  const updateSearchCount = () => {
    if (!searchText) {
      searchCount.style.display = 'none';
      return;
    }

    const counts = countMatches(activeRoot);
    searchCount.textContent = `${counts.matched} / ${counts.total} matched`;
    searchCount.style.display = 'inline-block';
    searchCount.className = counts.matched === 0 ? 'no-match' : '';
  };

  const render = () => {
    getLayout();
    const width = canvas.clientWidth;
    const height = canvas.clientHeight;
    ctx.clearRect(0, 0, width, height);
    ctx.globalAlpha = 1.0;
    drawNode(activeRoot, width, 0);

    if (!hoveredNode) return;

    const rect = getNodeRect(hoveredNode, width);
    if (!rect) return;
    ctx.strokeStyle = '#fff';
    ctx.lineWidth = 2;
    ctx.strokeRect(rect.x, rect.y, rect.w, rect.h);
  };

  // Draw only frames that are large enough and inside the current canvas bounds.
  const drawNode = (node, canvasWidth, depth = 0) => {
    if (depth >= MAX_API_STACK_DEPTH) {
      if (node.children && node.children.length > 0) showMaxDepthWarning();
      return;
    }

    const rect = getNodeRect(node, canvasWidth);
    if (!rect) return;

    const canvasHeight = canvas.clientHeight;
    if (rect.x + rect.w < 0 || rect.x > canvasWidth) return;
    if (rect.y + rect.h < 0 || rect.y > canvasHeight) return;
    if (rect.w < 0.5) return;

    const isAncestorNode = isAncestor(node);
    ctx.globalAlpha = isAncestorNode ? ANCESTOR_OPACITY : 1.0;
    ctx.fillStyle = getColor(node);
    ctx.fillRect(rect.x, rect.y, rect.w, rect.h);

    if (searchText && !matchesSearch(node)) {
      ctx.fillStyle = 'rgba(0,0,0,0.5)';
      ctx.fillRect(rect.x, rect.y, rect.w, rect.h);
    }

    ctx.strokeStyle = '#1e1e1e';
    ctx.lineWidth = 0.5;
    ctx.strokeRect(rect.x, rect.y, rect.w, rect.h);

    if (rect.w >= MIN_LABEL_WIDTH) {
      let label = node.name;
      while (label.length > 3 && ctx.measureText(`${label}...`).width > rect.w - 8) {
        label = label.slice(0, -1);
      }
      if (label.length < node.name.length) label += '...';
      ctx.fillStyle = '#fff';
      ctx.font = `${FONT_SIZE}px monospace`;
      ctx.textBaseline = 'middle';
      ctx.fillText(label, rect.x + 4, rect.y + rect.h / 2);
    }

    if (isAncestorNode) {
      const pathIndex = focusPath.indexOf(node);
      if (pathIndex >= 0 && pathIndex < focusPath.length - 1) {
        drawNode(focusPath[pathIndex + 1], canvasWidth, depth + 1);
      }
      return;
    }

    if (!node.children) return;
    for (const child of node.children) {
      drawNode(child, canvasWidth, depth + 1);
    }
  };

  // Convert normalized layout coordinates into canvas pixels, with depth 0 at the bottom.
  const getNodeRect = (node, canvasWidth) => {
    if (node._w === undefined || node._w < 1e-12) return null;
    const x = node._x * canvasWidth;
    const width = Math.max(node._w * canvasWidth, 1);
    const canvasHeight = canvas.clientHeight;
    return { x, y: canvasHeight - (node._d + 1) * ROW_HEIGHT, w: width, h: ROW_HEIGHT - 1 };
  };

  // Restrict hit testing to the target depth, then follow the focused path when zoomed in.
  const findNodeAt = (x, y, root, canvasWidth) => {
    const canvasHeight = canvas.clientHeight;
    const targetDepth = Math.floor((canvasHeight - y) / ROW_HEIGHT);

    const search = (node, depth = 0) => {
      if (depth >= MAX_API_STACK_DEPTH) {
        if (node.children && node.children.length > 0) showMaxDepthWarning();
        return null;
      }
      if (node._d > targetDepth) return null;
      if (node._d === targetDepth) {
        const rect = getNodeRect(node, canvasWidth);
        return rect && x >= rect.x && x <= rect.x + rect.w ? node : null;
      }

      const pathIndex = focusPath.indexOf(node);
      if (pathIndex >= 0 && pathIndex < focusPath.length - 1) {
        return search(focusPath[pathIndex + 1], depth + 1);
      }

      if (!node.children) return null;
      for (const child of node.children) {
        const result = search(child, depth + 1);
        if (result) return result;
      }
      return null;
    };

    return search(root);
  };

  const formatDuration = (us) => {
    if (us >= 1000000) return `${(us / 1000000).toFixed(2)} s`;
    if (us >= 1000) return `${(us / 1000).toFixed(2)} ms`;
    return `${us.toFixed(2)} us`;
  };

  const formatPercent = (value, total) => {
    if (total === 0) return '0%';
    return `${((value / total) * 100).toFixed(2)}%`;
  };

  const showTooltip = (node, cx, cy) => {
    tooltip.innerHTML = `
      <div class="fname">${escapeHtml(node.name)}<span class="cat cat-${node.category}">${node.category}</span></div>
      <div class="detail">Calls: ${node.count.toLocaleString()}</div>
      <div class="detail">Total: ${formatDuration(node.value)} (${formatPercent(node.value, totalDuration)})</div>
      <div class="detail">Self: ${formatDuration(node.self_time)} (${formatPercent(node.self_time, totalDuration)})</div>
    `;
    tooltip.style.display = 'block';

    let tx = cx + 12;
    let ty = cy - 10;
    if (tx + 300 > window.innerWidth) tx = cx - 300;
    if (ty + 100 > window.innerHeight) ty = cy - 100;
    tooltip.style.left = `${tx}px`;
    tooltip.style.top = `${ty}px`;
  };

  const escapeHtml = (text) => text
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;');

  canvas.addEventListener('mousemove', (event) => {
    const rect = canvas.getBoundingClientRect();
    const node = findNodeAt(
      event.clientX - rect.left,
      event.clientY - rect.top,
      activeRoot,
      canvas.clientWidth,
    );

    if (node) {
      hoveredNode = node;
      canvas.style.cursor = 'pointer';
      showTooltip(node, event.clientX, event.clientY);
    } else {
      hoveredNode = null;
      canvas.style.cursor = 'default';
      tooltip.style.display = 'none';
    }
    scheduleRender();
  });

  canvas.addEventListener('mouseleave', () => {
    hoveredNode = null;
    tooltip.style.display = 'none';
    scheduleRender();
  });

  canvas.addEventListener('click', (event) => {
    const rect = canvas.getBoundingClientRect();
    const node = findNodeAt(
      event.clientX - rect.left,
      event.clientY - rect.top,
      activeRoot,
      canvas.clientWidth,
    );
    if (!node) return;

    if (node === activeRoot) {
      focusPath = [activeRoot];
      layoutCache = null;
      resize(true);
      return;
    }

    const ancestorIndex = focusPath.slice(0, -1).indexOf(node);
    if (ancestorIndex >= 0) {
      focusPath = focusPath.slice(0, ancestorIndex + 1);
      layoutCache = null;
      resize(true);
      return;
    }

    focusPath = buildPath(node);
    layoutCache = null;
    resize(true);
  });

  // Rebuild the zoom path from parent links instead of scanning the full tree.
  const buildPath = (target) => {
    const path = [];
    let node = target;

    while (node) {
      path.push(node);
      if (node === activeRoot) break;
      node = node._parent;
    }

    if (path[path.length - 1] !== activeRoot) return [activeRoot];
    return path.reverse();
  };

  threadDropdownToggle.addEventListener('click', (event) => {
    event.stopPropagation();
    threadDropdown.classList.toggle('open');
  });

  document.addEventListener('click', (event) => {
    if (!threadDropdown.contains(event.target)) {
      threadDropdown.classList.remove('open');
    }
  });

  resetBtn.addEventListener('click', () => {
    focusPath = [activeRoot];
    layoutCache = null;
    resize(true);
  });

  searchInput.addEventListener('input', () => {
    searchText = searchInput.value;
    updateSearchCount();
    scheduleRender();
  });

  resetSearchBtn.addEventListener('click', () => {
    searchInput.value = '';
    searchText = '';
    updateSearchCount();
    scheduleRender();
  });

  document.addEventListener('keydown', (event) => {
    if (event.key !== 'Escape') return;

    if (searchText) {
      searchInput.value = '';
      searchText = '';
      updateSearchCount();
      scheduleRender();
    } else {
      resetBtn.click();
    }
  });

  window.addEventListener('resize', () => resize(true));

  initParentLinks(allRoot);
  populateThreadDropdown();
  updateThreadDropdownActive('__all__');
  resize(true);
})();

File: scripts/flame_graph/flamegraph.py
#!/usr/bin/env python3
"""
-------------------------------------------------------------------------
This file is part of the MindStudio project.
Copyright (c) 2026 Huawei Technologies Co.,Ltd.

MindStudio is licensed under Mulan PSL v2.
You can use this software according to the terms and conditions of the Mulan PSL v2.
You may obtain a copy of Mulan PSL v2 at:

         http://license.coscl.org.cn/MulanPSL2

THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
See the Mulan PSL v2 for more details.
-------------------------------------------------------------------------

Standalone Python script to extract PyTorch function call stacks from
Ascend PyTorch Profiler DB files and generate interactive flame graphs.

Usage:
    python flamegraph.py <db_path> [--output OUTPUT]
"""

import argparse
import html
import json
import logging
import os
import re
import sqlite3
import sys
from datetime import datetime
from typing import Dict, Iterator, List, NoReturn, Optional, Tuple


ApiCall = Tuple[int, int, Optional[str]]
ThreadInfo = Tuple[int, int, int]
BuildStats = Dict[str, int]


logger = logging.getLogger(__name__)


def _exit_with_error(message: str) -> NoReturn:
    logger.error(message)
    sys.exit(1)


def _validate_readable_file(file_path: str, description: str) -> str:
    path: str = os.path.abspath(file_path)
    if not os.path.exists(path):
        _exit_with_error(f"{description} does not exist: {path}")
    if not os.path.isfile(path):
        _exit_with_error(f"{description} is not a file: {path}")
    if not os.access(path, os.R_OK):
        _exit_with_error(f"{description} is not readable: {path}")
    return path


REQUIRED_TABLES = ["PYTORCH_API", "STRING_IDS"]

QUERY_API_CALLS_FILTERED = """
SELECT
    CAST(api.startNs AS INTEGER) AS startNs,
    CAST(api.endNs AS INTEGER) AS endNs,
    api.globalTid,
    api_name.value AS api_name
FROM PYTORCH_API AS api
LEFT JOIN STRING_IDS AS api_name ON api.name = api_name.id
WHERE api.startNs IS NOT NULL
  AND api.endNs IS NOT NULL
  AND api.globalTid = ?
  AND (api_name.value IS NULL OR api_name.value NOT LIKE 'ProfilerStep#%')
ORDER BY api.startNs ASC
"""

QUERY_THREADS = """
SELECT DISTINCT globalTid,
       globalTid >> 32 AS pid,
       globalTid & 0xFFFFFFFF AS tid
FROM PYTORCH_API
ORDER BY globalTid
"""

FETCH_SIZE = 10000
MAX_DB_FILE_SIZE = 10 * 1024 * 1024 * 1024
MAX_API_STACK_DEPTH = 1000


class ProfilerDBReader:
    """Reads PyTorch profiler DB and extracts call stacks with timing data."""

    def __init__(self, db_path: str) -> None:
        self.db_path: str = db_path
        self._conn: Optional[sqlite3.Connection] = None
        self._validate_and_connect()

    def close(self) -> None:
        if self._conn is not None:
            self._conn.close()
            self._conn = None

    def _validate_and_connect(self) -> None:
        self.db_path = _validate_readable_file(self.db_path, "DB file")
        if not self.db_path.lower().endswith(".db"):
            _exit_with_error(f"DB file must have .db extension, got: {self.db_path}")
        db_file_size = os.path.getsize(self.db_path)
        if db_file_size > MAX_DB_FILE_SIZE:
            _exit_with_error(f"DB file exceeds the 10GB size limit: {self.db_path} ({db_file_size / (1024**3):.2f} GB)")
        try:
            with open(self.db_path, "rb") as f:
                header: bytes = f.read(16)
        except OSError as err:
            _exit_with_error(f"Failed to read DB file header {self.db_path}: {err}")
        if not header.startswith(b"SQLite format 3\x00"):
            _exit_with_error(f"DB file is not a valid SQLite database: {self.db_path}")
        try:
            self._conn = sqlite3.connect(f"file:{self.db_path}?mode=ro", uri=True)
        except sqlite3.Error as err:
            _exit_with_error(f"Failed to open DB file {self.db_path}: {err}")
        try:
            cursor = self._conn.cursor()
            cursor.execute("SELECT name FROM sqlite_master WHERE type='table'")
            tables = {row[0] for row in cursor.fetchall()}
            missing = [t for t in REQUIRED_TABLES if t not in tables]
            if missing:
                _exit_with_error(f"DB is missing required tables: {', '.join(missing)}")
        except sqlite3.Error as err:
            _exit_with_error(f"Failed to validate DB file {self.db_path}: {err}")

    def list_threads(self) -> List[ThreadInfo]:
        """Return list of (globalTid, pid, tid) tuples."""
        try:
            cursor = self._conn.cursor()
            cursor.execute(QUERY_THREADS)
            return cursor.fetchall()
        except sqlite3.Error as err:
            _exit_with_error(f"Failed to query threads from DB file {self.db_path}: {err}")

    def read_api_calls(self, global_tid: int) -> Iterator[ApiCall]:
        """
        Yield (start_ns, end_ns, api_name) tuples for the given thread.

        Uses streaming read to handle large DBs efficiently.
        """
        try:
            cursor = self._conn.cursor()
            cursor.execute(QUERY_API_CALLS_FILTERED, (global_tid,))

            while True:
                rows = cursor.fetchmany(FETCH_SIZE)
                if not rows:
                    break
                for start_ns, end_ns, _, api_name in rows:
                    yield (start_ns, end_ns, api_name)
        except sqlite3.Error as err:
            _exit_with_error(f"Failed to query API calls from DB file {self.db_path}: {err}")


class FlameNode:
    """A node in the flame graph call tree."""

    __slots__ = ("name", "category", "value", "self_time", "count", "children")

    def __init__(self, name: str, category: str = "unknown") -> None:
        self.name: str = str(name)
        self.category: str = category
        self.value: float = 0.0
        self.self_time: float = 0.0
        self.count: int = 0
        self.children: Dict[str, FlameNode] = {}

    def _to_dict_node(self) -> Dict[str, object]:
        """Create a dict for this node only (no children)."""
        return {
            "name": self.name,
            "category": self.category,
            "value": round(self.value, 2),
            "self_time": round(self.self_time, 2),
            "count": self.count,
        }


def _to_dict_iterative(root: "FlameNode") -> Dict[str, object]:
    """Iterative serialization of FlameNode tree with a max API call stack depth guard."""
    result = root._to_dict_node()
    stack = [(root, result, 0)]
    truncated_nodes = 0
    while stack:
        node, parent_dict, depth = stack.pop()
        if not node.children:
            continue
        children_list = sorted(node.children.values(), key=lambda n: n.value, reverse=True)
        parent_dict["children"] = []
        if depth >= MAX_API_STACK_DEPTH:
            truncated_nodes += len(children_list)
            continue
        for child in children_list:
            child_dict = child._to_dict_node()
            parent_dict["children"].append(child_dict)
            stack.append((child, child_dict, depth + 1))
    if truncated_nodes:
        logger.warning(
            "Flame graph serialization reached the max API call stack depth %d; truncated %d child nodes.",
            MAX_API_STACK_DEPTH,
            truncated_nodes,
        )
    return result


FRAMEWORK_KEYWORDS_LOWER = ("torch", "torch_npu", "aten::", "c10::", "aten_")
CANN_KEYWORDS_LOWER = ("cann", "ascendcl", "aclnn", "aclrt", "aclmdl", "aclprof", "hccl")


def classify_frame(frame_text: str) -> str:
    """Classify a stack frame or API name into a category."""
    text_lower = frame_text.lower()
    for ckw in CANN_KEYWORDS_LOWER:
        if ckw in text_lower:
            return "cann"
    for fkw in FRAMEWORK_KEYWORDS_LOWER:
        if fkw in text_lower:
            return "python_framework"
    if ".py" in frame_text or "python" in text_lower:
        return "python"
    return "unknown"


def _build_single_thread_tree(reader: ProfilerDBReader, global_tid: int) -> Tuple[FlameNode, int, float, BuildStats]:
    """Build a call tree for a single thread and return (root, calls, duration, stats)."""
    root = FlameNode("all", "root")
    root_end_ns = float("inf")
    stack = [(root, root_end_ns)]
    total_calls = 0
    stats = {"non_positive_duration": 0, "empty_api_name": 0}

    for start_ns, end_ns, api_name in reader.read_api_calls(global_tid):
        duration_us = (end_ns - start_ns) / 1000.0
        total_calls += 1
        if duration_us <= 0:
            stats["non_positive_duration"] += 1
            continue
        if not api_name:
            stats["empty_api_name"] += 1
            api_name = "unknown"

        while len(stack) > 1 and stack[-1][1] <= start_ns:
            stack.pop()

        parent = stack[-1][0]
        category = classify_frame(api_name)
        if api_name not in parent.children:
            parent.children[api_name] = FlameNode(api_name, category)
        child = parent.children[api_name]
        child.value += duration_us
        child.count += 1
        for ancestor, _ in stack:
            ancestor.count += 1

        if end_ns > start_ns:
            stack.append((child, end_ns))

    _compute_self_time(root, is_root=True)
    return root, total_calls, root.value, stats


def _merge_trees(dst: FlameNode, src: FlameNode) -> None:
    """Merge src tree into dst tree, combining inclusive values and counts."""
    stack = [(dst, src)]
    while stack:
        dst_node, src_node = stack.pop()
        dst_node.value += src_node.value
        dst_node.count += src_node.count
        for name, src_child in src_node.children.items():
            if name not in dst_node.children:
                dst_node.children[name] = FlameNode(name, src_child.category)
            stack.append((dst_node.children[name], src_child))


def _compute_self_time(node: FlameNode, is_root: bool = False) -> None:
    """Recompute exclusive self time from inclusive value and children."""
    stack = [(node, False)]
    while stack:
        current, visited = stack.pop()
        if visited:
            children_value = sum(c.value for c in current.children.values())
            if current is node and is_root:
                current.value = children_value
                current.self_time = 0.0
            else:
                current.self_time = max(0.0, current.value - children_value)
            continue
        stack.append((current, True))
        for child in current.children.values():
            stack.append((child, False))


def build_flame_tree(reader: ProfilerDBReader) -> Tuple[FlameNode, int, float, BuildStats]:
    """
    Build a flame graph tree by processing each thread independently
    and merging the results.

    Returns (root_node, total_calls, total_duration_us, stats).
    """
    root = FlameNode("all", "root")
    total_calls = 0
    stats = {"non_positive_duration": 0, "empty_api_name": 0}

    threads = reader.list_threads()

    for global_tid, pid, tid in threads:
        thread_root, t_calls, _, thread_stats = _build_single_thread_tree(reader, global_tid)
        thread_key = f"thread_{tid} (globalTid: {global_tid})"
        if thread_key not in root.children:
            root.children[thread_key] = FlameNode(thread_key, "root")
        _merge_trees(root.children[thread_key], thread_root)
        total_calls += t_calls
        for key, value in thread_stats.items():
            stats[key] = stats.get(key, 0) + value

    root.value = sum(child.value for child in root.children.values())
    _compute_self_time(root)
    return root, total_calls, root.value, stats


SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))


def generate_html(
    root: FlameNode,
    total_calls: int,
    total_duration_us: float,
    title: str,
    output_path: str,
) -> None:
    """Generate a self-contained HTML flame graph file with inlined JS."""

    template_path = _validate_readable_file(
        os.path.join(SCRIPT_DIR, "flamegraph_template.html"),
        "HTML template file",
    )
    js_path = _validate_readable_file(
        os.path.join(SCRIPT_DIR, "flamegraph.js"),
        "JavaScript file",
    )

    try:
        with open(template_path, "r", encoding="utf-8") as f:
            template_text = f.read()
    except OSError as err:
        _exit_with_error(f"Failed to read HTML template file {template_path}: {err}")

    try:
        with open(js_path, "r", encoding="utf-8") as f:
            js_code = f.read()
    except OSError as err:
        _exit_with_error(f"Failed to read JavaScript file {js_path}: {err}")

    metadata = {
        "total_duration_us": round(total_duration_us, 2),
        "num_calls": total_calls,
        "generated_at": datetime.now().isoformat(),
    }
    data = {
        "metadata": metadata,
        "flamegraph": _to_dict_iterative(root),
    }
    # json.dumps with ensure_ascii=True escapes < as \u003c and > as \u003e,
    # which prevents HTML injection when JSON is inlined inside a script tag.
    json_str = json.dumps(data, ensure_ascii=True, separators=(",", ":"))
    # Only escape script-closing sequences in JS; escaping all angle brackets
    # would break JavaScript comparison operators such as "<" and ">".
    js_code_safe = re.sub(r"</script", r"<\\/script", js_code, flags=re.IGNORECASE)

    html_text = template_text.replace("$title", html.escape(title, quote=False))
    html_text = html_text.replace("$json_data", json_str)
    html_text = html_text.replace("$js_code", js_code_safe)

    try:
        with open(output_path, "w", encoding="utf-8") as f:
            f.write(html_text)
    except OSError as err:
        _exit_with_error(f"Failed to write output file {output_path}: {err}")

    try:
        size_mb = os.path.getsize(output_path) / (1024 * 1024)
    except OSError as err:
        _exit_with_error(f"Failed to inspect output file {output_path}: {err}")
    logger.info("Flame graph written to: %s (%.1f MB)", output_path, size_mb)


def main() -> None:
    logging.basicConfig(level=logging.INFO, format="%(message)s")
    parser = argparse.ArgumentParser(
        description="Extract PyTorch API call hierarchy from Ascend Profiler DB "
        "and generate an interactive flame graph.",
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    parser.add_argument(
        "db_path",
        nargs="?",
        help="Path to ascend_pytorch_profiler_{{Rank_ID}}.db file",
    )
    parser.add_argument(
        "--output",
        "-o",
        default=".",
        help="Output directory for flamegraph.html (default: current directory)",
    )
    args = parser.parse_args()

    if not args.db_path:
        parser.error("db_path is required")

    reader = ProfilerDBReader(args.db_path)
    try:
        logger.info("Reading: %s", args.db_path)

        logger.info("Building flame tree from API call intervals...")
        root, total_calls, total_duration_us, build_stats = build_flame_tree(reader)

        if total_calls == 0:
            logger.error("No API call data found in the DB.")
            sys.exit(1)

        skipped_calls = build_stats.get("non_positive_duration", 0)
        empty_names = build_stats.get("empty_api_name", 0)
        logger.info("Total: %s API calls, total duration: %.2f s", total_calls, total_duration_us / 1000000)
        if skipped_calls or empty_names:
            logger.info(
                "Invalid records: %s non-positive durations skipped, %s empty API names replaced with unknown",
                skipped_calls,
                empty_names,
            )

        title = os.path.basename(reader.db_path)

        output_dir = os.path.abspath(args.output)
        if not os.path.isdir(output_dir):
            _exit_with_error(f"Output path is not a directory: {output_dir}")
        if not os.access(output_dir, os.W_OK):
            _exit_with_error(f"Output directory is not writable: {output_dir}")
        output_path = os.path.join(output_dir, "flamegraph.html")

        logger.info("Generating flame graph...")
        generate_html(root, total_calls, total_duration_us, title, output_path)
        logger.info("Done! Open the HTML file in a browser to view.")
    finally:
        reader.close()


if __name__ == "__main__":
    main()
